plugins/adfs/adfs.py

```diff
@@ -0,0 +1,82 @@
+import datetime, requests
+from utils.utils import generate_ip, generate_id, generate_trace_id
+
+
+def adfs_authenticate(url, username, password, useragent, pluginargs):
+
+    ts = datetime.datetime.utcnow().strftime('%Y-%m-%d %H:%M:%S')
+
+    data_response = {
+        'timestamp': ts,
+        'username': username,
+        'password': password,
+        'success': False,
+        'change': False,
+        '2fa_enabled': False,
+        'type': None,
+        'code': None,
+        'name': None,
+        'action': None,
+        'headers': [],
+        'cookies': [],
+        'sourceip': None,
+        'throttled': False,
+        'error': False,
+        'output': ""
+    }
+
+    # post_data = urllib.parse.urlencode({'UserName': username, 'Password': password,
+    #                                    'AuthMethod': 'FormsAuthentication'}).encode('ascii')
+    post_data = {
+        'UserName': username,
+        'Password': password,
+        'AuthMethod': 'FormsAuthentication'
+    }
+
+    # ?client-request-id=&wa=wsignin1.0&wtrealm=urn:federation:MicrosoftOnline&wctx=cbcxt=&username={}&mkt=&lc=
+    params_data =  {
+        'client-request-id' : '',
+        'wa' : 'wsignin1.0',
+        'wtrealm' : 'urn:federation:MicrosoftOnline',
+        'wctx' : '',
+        'cbcxt' : '',
+        'username' : username,
+        'mkt' : '',
+        'lc' : ''
+    }
+
+    spoofed_ip = generate_ip()  # maybe use client related IP address
+    amazon_id = generate_id()
+    trace_id = generate_trace_id()
+
+    headers = {
+        'User-Agent': 'Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:65.0) Gecko/20100101 Firefox/65.0', #useragent,
```

Comment:
user agent should be the input, not static

Reply:
Absolutely. Provided the change a few minutes ago. Note: the CredMaster's default user agent did not work for my ADFS targets. This is why I changed it but I will give this hint in the wiki documentation then instead.